### ISSUE TITLE
rdadmin sql error when creating DB (v2.13.0)

### ISSUE THREAD

**Issue by waynemerricks** [ElvishArtisan/rivendell#121]
This is different to the MyISAM errors you can fix that by setting a default engine
in my.cnf and restarting the service.

I'm running MySQL ver 14.14 Distrib 5.7.12 on Ubuntu 16.04

When you first run rdadmin to create the DB you get this error:

> SQL: insert into LOGS 
> (NAME,SERVICE,DESCRIPTION,ORIGIN_USER,ORIGIN_DATETIME,LINK_DATETIME)
> values ("SAMPLE","Production","Sample Log","user",NOW())
> SQL Error: Field 'LINK_DATETIME' doesn't have a default value

Looking in createdb.cpp on line 1231 you create the table as follows:

> sql=QString("CREATE TABLE IF NOT EXISTS LOGS (\
> ...
>   LINK_DATETIME DATETIME NOT NULL
> ...

It is actually an issue with all of these fields:

LINK_DATETIME
MODIFIED_DATETIME
START_DATE
END_DATE

So you need to change line 2646 on createdb.cpp to include several more NOW values:

`sql="insert into LOGS (NAME,SERVICE,DESCRIPTION,ORIGIN_USER,ORIGIN_DATETIME)\
 values (\"SAMPLE\",\"Production\",\"Sample Log\",\"user\",NOW())";`

becomes:

`sql="insert into LOGS (NAME,SERVICE,DESCRIPTION,ORIGIN_USER,ORIGIN_DATETIME,LINK_DATETIME,
 MODIFIED_DATETIME,START_DATE,END_DATE) values (\"SAMPLE\",\"Production\",\"Sample Log\",\"user\",NOW(),NOW(),NOW(),NOW(),NOW())";`

**Comment by waynemerricks** [comment]
As a further note, once you get past this error by fixing the source, rdadmin freezes the entire machine when you reboot all the DB tables are corrupt.

I'm not sure what is going on with that but if I import a backup from my production server (Riv 2.10.3), this same install will upgrade to 2.13.0 and works as normal so I don't think it is because I messed up the QT driver or anything.

**Comment by waynemerricks** [comment]
Patch file [here](http://thevoiceasia.com/rivendell/createdb.cpp.patch)

**Comment by ElvishArtisan** [comment]
Interesting.  I am not seeing the error here (CentOS 7.2 / MariaDB 5.5.47).  I do see an issue, but it is actually with with column definitions for the LINK_DATETIME, START_DATE and END_DATE fields in the LOGS table.  NULL values here should be no problem.

**Comment by ElvishArtisan** [comment]
After a bit of research and testing, it would appear that many [earlier?] versions of MySQL and MariaDB do 'interesting' things with NULLs in date/time fields.  On MariaDB 5.5.47, attempting to write NULL to a datetime field with the 'not null' attribute does not generate an error, but actually results in an entry of "0000-00-00 00:00:00" - perhaps 'null' in the 'date/time' sense, but certainly not in the SQL one!  After removing the 'not null' attribute however, writing a NULL is properly reflected in the resulting value.  Even more interesting, if you then apply an ALTER TABLE statement to restore the 'not null' attribute, those NULLs are then converted into '0000-00-00 00:00:00' as part of the table change!  So it would appear that this behavior is [was] very intentional.

Regardless, it's a hack, and I've just committed a change to stable that removes the 'not null' attributes from LINK_DATETIME, START_DATE and END_DATE fields in the LOGS table, which should now be compatible with the stricter interpretation.  I've also added a NOW() default value for the MODIFIED_DATETIME field when creating the SAMPLE log, as that field retains its not null attribute.  

Please test!  (Heads Up: the change is implemented as a schema update, so be prepared accordingly).

**Comment by waynemerricks** [comment]
This along with the create user changes resolves the db creation issues.

Cheers.

**Comment by waynemerricks** [comment]
Right, I've been on a multi day trip grep'ing through code and testing results.  I've found some random errors related to the date changes (and some just plain errors).  I'll post them here to aid my memory and then try to work on patches as necessary after that.

Please note, this is not an exhaustive list.  I tried to find as much as possible but I've no doubt missed bits (I started with a 22,000 line file of grep results).  Also, line numbers are for Riv 2.13.0 as thats what I'm trying to get working.

 - lib/rdevent_line.cpp:568
 -- Missing Field 'CART_NUMBER' from the insert into 'X_LOG' statement

 - lib/rdreport.cpp:554
 -- Warning here, you are inserting into 'X_SRT' but adding a full date time to a time only column (e.g. yyyy-mm-dd hh:mm:ss instead of just hh:mm:ss)

 - lib/rdsvc.cpp:624
 -- There is an insert into %s statement but I can't find the table that these columns belong.  Is this just some old cruft hanging around?
 
 - lib/rdsvc.cpp:768
 -- Update logs here causes issues with START_DATE and END_DATE being null (this is fixed with your table changes)
 
 - lib/rdsvc.cpp:789
 -- This is another insert into LOGS.  There are two cases, if PURGE_DATE is set (you have a PURGE set on the LOGS table, e.g. purge after 30 days) this is fine.  But if PURGE_DATE is not set (the table is -1), then you have an empty "QString purge_date;" with no initialised value.  I'm actually not sure if this would crash Riv (need to test) but at the very least you get an "incorrect date value '' " SQL error.
 
 - rdlogedit/rdlogedit.cpp:472
 -- Standard insert into logs suffering from the no defaults but not null on these fields, MODIFIED_DATETIME, START_DATE, END_DATE, PURGE_DATE and IMPORT_DATE
 
 - lib/rdlog.cpp:46
 -- Again an insert and these same fields plus one extra; SERVICE, START_DATE, END_DATE, PURGE_DATE and IMPORT_DATE

Before I start hacking patches together, is there a problem with setting a default or NULL on these fields?

**Comment by waynemerricks** [comment]
Back again, most of these are trivial fixes but rdlog isn't.
- lib/rdevent_line.cpp:568: CART_NUMBER is not accessible in this section of code.  This is the Import links bit of inserting into X_LOG.  Before 568, you get the event_length by looping through all the carts and adding the length up so it doesn't seem to apply to any one CART_NUMBER.  
 
 So I set CART_NUMBER to 0 which seems to be valid for some other entries I have in my X_LOG tables.

 - lib/rdreport.cpp:554: I changed the toString to hh:mm:ss instead of the full date time
 - lib/rdsvc.cpp:789: I did an isEmpty check on purge_date and removed it from the SQL if it was empty.
 - rdlogedit/rdlogedit.cpp:472: Added MODIFIED_DATETIME=NOW() to the SQL

Now the annoying fix; rdlog.  This class, if you supply it with a name and set create to true it will insert a new log if it doesn't already exist.  If you don't specify create, it just initialises the object.  The problem is to initialise the object you need a service name if you intend to create a log.

I changed the constructor definition of RDLog to separate out log creation.  There is now a simple RDLog(logname) for no creation and RDLog(logname, servicename) for creation (this replaced the logname, create definition).

This impacts two places:
 - rdairplay/list_log.cpp:898: Just need to add svcname instead of true to create log no problem here
 - rdlogedit/edit_log.cpp:1092: same fix as above

I need to test these changes in anger so bear with me while I go through everything.

**Comment by waynemerricks** [comment]
grep searches and db alterations didn't fix Manuel's problem.  Recreating the problem is the easy part, unfortunately the terminal doesn't give me a nice stack trace just a list of SQL statements that error (one day I'll get Eclipse going with QT3, QTCreator seems fairly useless).

Open a log, add or modify an event.  OK out of it and see the terminal errors.  Open the log again and the changes were not saved.

You get errors for setting PURGE_DATE, START_DATE, END_DATE and ORIGIN_DATETIME to "" in different code places.

It partially boils down to the rdlog_line.cpp file.  In the constructor you call clear which sets log_start_time, log_origin_datetime, log_year, log_start_datetime, log_end_datetime, log_play_time, log_marker_post_time, log_ext_start_time and log_link_start_time to either a QDate(), QDateTime() or QTime() respectively.

The problem being these are all null times and when you do a toString on these, you get "".  The only one causing an issue here is log_origin_datetime but in principal the other values might cause issues elsewhere in Riv.

I can't figure out where PURGE_DATE, START_DATE and END_DATE are set as it seems to be in one of the helper classes that do this kind of thing "UPDATE %s SET %s=%s" so it is impossible to fix the bug there.

The alternative is to alter RDLogLine to return valid dates and times or initially set them with Clear to something usable by MySQL.  If I change the clear statement for originDateTime to QDateTime::currentDateTime() it clears the INSERT error so at least your events are stored but obviously the errors for PURGE, START and END still persist.

**Comment by ElvishArtisan** [comment]
Re: CART_NUMBER in the log tables: instead of writing zeros, the cleaner fix is to set the default value of the field to '0'.

Just pushed to stable as DB version 258.

**Comment by ElvishArtisan** [comment]
Perhaps an RDSqlDateTime() wrapper to handle the NULL conversions?  Something like:
<code>
```
QString RDSqlDateTime(const QDate &date)
{
    if(date.isValid()) {
      return "\""+date.toString("yyyy-MM-dd")+"\"";
    }
    return "NULL";
}
```
</code>
You'd use it like:
<code>
```
sql="update TABLE set SOME_DATE="+RDSqlDateTime(my_date);
```
</code>
I'm thinking that 'lib/rdescape_string.[cpp,h] would be the place to put it.  You would of course need overloaded versions that take a QTime and a QDateTime as well.

**Comment by waynemerricks** [comment]
Seems like the best way of doing it but then you'd need to modify every part of Riv that was using dates/times?  You could also just change any SetRow method you found to include the date changes, again might be a lot of work.

I found purge date etc, it is part of setStart/End/PurgeDate in lib/rdlog.cpp.  I can fix those errors quite easily by changing:

void RDLog::setPurgeDate(const QDate &date) const
{
  SetRow("PURGE_DATE",date);
}

to

void RDLog::setPurgeDate(const QDate &date) const
{
  if(date.isValid())
    SetRow("PURGE_DATE",date);
  else
    SetRow("PURGE_DATE", "NULL");
}

**Comment by ElvishArtisan** [comment]
Yes, it would be a lot of work, but I suspect we'd be better off doing that then trying to nit-pick the individual situations where it's needed.  Do it once, wrapping every date/time value where it's used in a SQL statement, and then we're done with it.

**Comment by waynemerricks** [comment]
This is actually not as bad as I thought it would be.  A quick grep through of all the QDate/QTime and QDateTime references, excluding header files, I have 859 lines of code to check (most of them aren't directly SQL related but my grep powers don't extend far enough to auto filter those out).

I'll get back to you once I've narrowed the 859 down.

**Comment by waynemerricks** [comment]
OK, after thinking about this a bit more I realised I had more things to search for.  I based this on the stable branch as of yesterday.

I grepped for QDate, QTime, time string variations (e.g. yyyy-mm-dd hh:mm:ss) and all the database fields I could find that had date/time/timestamp fields.  That gave me 1,592 lines to check.  A quick look through those one by one and I'm left with 248 lines of code in various parts of Riv that might need modification (some of these are calling the same function so multiples can be fixed with one function tweak).

Oh, I also found a bug: rdevent_line.cpp:415.  You're doing an insert into EVENTS table with a where clause at the end.  This seems to be a copy/paste of previous code but the where clause and associated variable should have been removed.

There is also some code looking at *_CLK tables but I'm guessing that is only there because stable is migrating the clock tables to the CLOCKS and CLOCK_METADATA tables (and is not quite done yet).

I'll try and start putting up pull requests but it will take a few days to organise everything.
